lib/DojotHandler.js

```diff
@@ -38,91 +39,109 @@ module.exports = class DojotHandler {
       return;
     }
 
-    this.commandSocket = zmq.socket('rep');
+    this.sock = zmq.socket('router');
 
-    this.commandSocket.on("message", (request) => {
-      let data;
+    this.sock.on("message", (identity, packet) => {
+      let request;
+      //parse the packet
       try {
-        data = JSON.parse(request.toString());
+        request = JSON.parse(packet);
       } catch (error) {
-        return this._handleResults(new Error("Payload must be valid json"));
+        console.log("Invalid JSON format. Discarding request: %s. Error:",
+          packet.toString(), error);
+        return;
       }
-
       try {
-        checkField(data, 'command', "Request is missing command field");
-        console.log('Got message. invoking handler ...', data);
-        switch (data.command) {
-          case 'locale':
-            this._handleLocale(data);
-            break;
-          case 'metadata':
-            this._handleMetadata(data);
-            break;
-          case 'message':
-            this._handleMessage(data);
-            break;
-          case 'html':
-            this._handleHtml(data);
-            break;
-          default:
-            this._handleResults(new Error("Unknown command requested"));
-        }
+        // console.log('Got message. invoking handler ...', payload);
+        checkField(request, 'requestId', "Request is missing requestId field");
+        checkField(request, 'payload', "Request is missing payload field");
+
+        this._handleRequest(request.payload).then((response) => {
+          let responsePacket = {
+            payload: response,
+            requestId: request.requestId
+          };
+          this.sock.send([identity, JSON.stringify(responsePacket)]);
+        }).catch( (error) => {
+          let responsePacket = {
+            payload: { error },
+            requestId: request.requestId
+          };
+          this.sock.send([identity, JSON.stringify(responsePacket)]);
+        });
       } catch (error) {
-        return this._handleResults(error);
+        console.log("Exception: " + error);
```

Comment:
Should an answer be sent back to flowbroker?

Reply:
Agree, I will change it.

Reply:
Thought that errors in node implementation (executed by handleMessage) would not be sent back to flowbroker.